[SYCL][ESIMD] Add test on simd bitwise not operator

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add test on simd bitwise not operator

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+  }
+};
+
+// Replace specific reference values to the bigger ones, so we can safely
+// substract `m_value` later.
```
There is no substraction operator here

vasilytric replied:
No, the `For_bitwise_not` has been updated in this commit: c43b67ab29daf8efbe8da054c33421dd1b03050f

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+// Replace specific reference values to the bigger ones, so we can safely
+// substract `m_value` later.
+template <typename T> class For_bitwise_not {
+  T m_value;
```
Do we need it?

vasilytric replied:
No, the `For_bitwise_not` has been updated in this commit: c43b67ab29daf8efbe8da054c33421dd1b03050f

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+public:
+  For_bitwise_not() = default;
+
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T>) {
```
Please add static_assert to fail on floating point types (state and validate preconditions explicitly)

vasilytric replied:
`static_assert` has been added in [here](https://github.com/intel/llvm-test-suite/pull/850/commits/c43b67ab29daf8efbe8da054c33421dd1b03050f#diff-bb4bf7952553e3ab4d03fc650879928067a4565d9c86741e8c0344e689c0b04eR55-R56)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T>) {
+      if (std::abs(value<T>::lowest() + 1) == (value<T>::max() - 1)) {
+        if (val == value<T>::max()) {
+          val -= 1;
+        } else if (val == 0) {
+          val = 1;
+        }
```
Would you mind to add comments clarifying which branch is required for one's complement, and which branch is required for signed magnitude representation?

vasilytric replied:
The comment has been added in this commit: c43b67ab29daf8efbe8da054c33421dd1b03050f

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+
+  void operator()(T &val) {
+    if constexpr (std::is_signed_v<T>) {
+      if (std::abs(value<T>::lowest() + 1) == (value<T>::max() - 1)) {
```
```suggestion
      if (std::abs(value<T>::lowest() + 1) == (value<T>::max() - 1)) {
      // We are not using two's complement signed number representation
```
On the second thought, even C++20 having only two's complement representation still has UB based on real-world compiler optimizations for signed integer value overflow. Could you possibly double-check if this condition is required? My concern is possibility of compiler optimizations in case of max/min usage for two's complement signed number representation.

vasilytric replied:
The `For_bitwise_not` has been updated in this commit: c43b67ab29daf8efbe8da054c33421dd1b03050f

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+
+  bool passed = true;
+
+  const auto uint_types = get_tested_types<tested_types::uint>();
```
```suggestion
  // Running test for all integral types
  const auto uint_types = get_tested_types<tested_types::uint>();
```

vasilytric replied:
Same comment has been added [here](https://github.com/intel/llvm-test-suite/pull/850/commits/66e0d60d08be20c2069e90e626ef302a7446303a#diff-438d222b55125297a312d73be031ed90355f82de7eee316af778345a1c741a0dR43) and [here](https://github.com/intel/llvm-test-suite/pull/850/commits/66e0d60d08be20c2069e90e626ef302a7446303a#diff-c2cae8f61bdba7b88270c263569fede5cea9bb1fea2c721d60c10f9407b16f41R37)

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.cpp:
```diff
@@ -0,0 +1,176 @@
+          val = 1;
+        }
+      }
+    }
```
```suggestion
    if constexpr (std::is_signed_v<T>) {
      // We could have UB for negative zero in different integral type
      // representations: two's complement, ones' complement and signed
      // magnitude.
      // See http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2218.htm
      // Note that there is no check for UB with padding bits here, as it would
      // effectively disable any possible check for signed integer bitwise
      // operations.

      static const T max = value<T>::max();
      static const T lowest = value<T>::lowest();

      if (std::abs(lowest + 1) == (max - 1)) {
        // C11 standard mentions that it's possible to have a `0b100...0` value
        // as a trap value for twos' complement representation. In such case the
        // condition above would trigger for twos' complement representation
        // also.

        if (val == max) {
          // Would result in trap representation for signed magnitude 
          val -= 1;
        } else if (val == 0) {
          // Would result in trap representation for ones' complement
          val = 1;
        }
      }
    } //  signed integral types
```
Based on C++17 [basic.fundamental] and C11 "Integer types"

vasilytric replied:
This comment has been applied in this commit: https://github.com/intel/llvm-test-suite/pull/850/commits/c43b67ab29daf8efbe8da054c33421dd1b03050f#diff-bb4bf7952553e3ab4d03fc650879928067a4565d9c86741e8c0344e689c0b04eR55-R56

yuriykoch reviewed (commented):
Actually, I'm not quite sure it's safe to use bitwise operators over signed integral values in general.

The reasons behind are:

- general rule:  [INT13-C](https://wiki.sei.cmu.edu/confluence/display/c/INT13-C.+Use+bitwise+operators+only+on+unsigned+operands)
- trap values and padding bits possibility: [n2218](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2218.htm) mentions all details
- possible existence of compiler optimizations relying on such restrictions (like the ones for arithmetical operations mentioned in [cppcon 2018](https://www.youtube.com/watch?v=JhUxIVf1qok))

Tagging @kbobrovs for awareness

yuriykoch commented:
Looks like while there is possibility of trap values and compiler optimizations in general C++17, the llvm itself states that signed integer types are guaranteed to be two's complement:

- https://llvm.org/docs/LangRef.html
- https://bugs.llvm.org/show_bug.cgi?id=950

Given that information I would suggest to do the following:

1. provide coverage for signed integral types, making it as generic as possible from the point of C++17 spec (using `For_bitwise_not`),
2. but split `operator_bitwise_not.cpp` into the `operator_bitwise_not_uint.cpp` and `operator_bitwise_not_sint.cpp` to have specific test on signed integral types;
3. mention all links from the discussion in `operator_bitwise_not_sint.cpp` description to make the future support easier

@vasilytric Feel free to state any concerns for such solution. The other possible option is to make a two's complement representation a test precondition (run `assert` in `operator_bitwise_not_sint.cpp`, so there would be no need in `For_bitwise_not` in such case)

vasilytric committed:
[SYCL][ESIMD] Update For_bitwise_not

- Add comments about possible UB
- Remove unnecessary comment
- Remove Unnecessary field

vasilytric committed:
[SYCL][ESIMD] Split test into sint and uint

Because we have possibility of trap values and compiler optimizations
in general C++17.

vasilytric committed:
Merge branch 'intel' into bitwise_not

vasilytric committed:
[SYCL][ESIMD] Use new ways to log messages

vasilytric committed:
[SYCL][ESIMD] Update operator_bitwise_not_sint

- Update comment
- use operators namespace

vasilytric commented:
> Looks like while there is possibility of trap values and compiler optimizations in general C++17, the llvm itself states that signed integer types are guaranteed to be two's complement:


The test has been split in two in this commit: https://github.com/intel/llvm-test-suite/pull/850/commits/66e0d60d08be20c2069e90e626ef302a7446303a

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.hpp:
```diff
@@ -0,0 +1,171 @@
+                            DataT *const operator_result) {
+    auto simd_obj = esimd::simd<DataT, NumElems>();
+    simd_obj.copy_from(ref_data);
+    auto bitwise_not_result = ~simd_obj;
```
Nit
```suggestion
    const auto bitwise_not_result = ~simd_obj;
```

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.hpp:
```diff
@@ -0,0 +1,171 @@
+    for (size_t i = 0; i < NumElems; ++i) {
+      {
+        DataT retrieved = source_simd_result[i];
+        DataT expected = ref_data[i];
```
Nit
```suggestion
        const DataT& retrieved = source_simd_result[i];
        const DataT& expected = ref_data[i];
```

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.hpp:
```diff
@@ -0,0 +1,171 @@
+        if (!are_bitwise_equal(expected, retrieved)) {
+          passed = false;
+          log::fail(TestDescriptionT(data_type), "Unexpected value at index ",
+                    i, ", retrieved: ", retrieved, ", expected: ", expected);
```
```suggestion
          log::fail(TestDescriptionT(data_type), "Unexpected source simd value at index ",
                    i, ", retrieved: ", retrieved, ", expected: ", expected);
```

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.hpp:
```diff
@@ -0,0 +1,171 @@
+        DataT expected = ~shared_ref_data[i];
+        if (!are_bitwise_equal(expected, retrieved)) {
+          passed = false;
+          log::fail(TestDescriptionT(data_type), "Unexpected value at index ",
+                    i, ", retrieved: ", retrieved, ", expected: ", expected);
```
```suggestion
          log::fail(TestDescriptionT(data_type), "Unexpected result value at index ",
                    i, ", retrieved: ", retrieved, ", expected: ", expected);
```

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.hpp:
```diff
@@ -0,0 +1,171 @@
+      log::print_line("Test failed due to type of the object that returns " +
+                      TestCaseT::get_description() +
+                      " operator is not equal to the expected one for simd<" +
+                      data_type + ", " + std::to_string(NumElems) + ">.");
```
```suggestion
      passed = false;
      log::fail(TestDescriptionT(data_type), "Invalid return type for operator");
```
Consider to use `log::fail` for failure logs. The message itself is up to you

yuriykoch commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not_sint.cpp:
```diff
@@ -0,0 +1,51 @@
+// not operator.
+// The test verifies that data from simd is not corrupted after calling bitwise
+// not operator, that bitwise not operator return type is as expected and
+// bitwise not operator result values is correct.
```
```suggestion
// bitwise not operator result values are correct.
```
Here and in `*_uint.cpp` file

vasilytric committed:
[SYCL][ESIMD] Small updates

- Make some variables const
- Update log messages
- Update some comments

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

vasilytric commented:
@v-klochkov Would you mind to review this PR?

v-klochkov commented on SYCL/ESIMD/api/functional/operators/operator_bitwise_not.hpp:
```diff
@@ -0,0 +1,171 @@
+        if (!are_bitwise_equal(expected, retrieved)) {
+          passed = false;
+          log::fail(TestDescriptionT(data_type),
+                    "Unexpected source simd value value at index ", i,
```
```suggestion
                    "Unexpected source simd value at index ", i,
```

vasilytric replied:
This comment has been applied in this commit: https://github.com/intel/llvm-test-suite/pull/850/commits/20cf48a13e17bd67c9b4586af860388fa7d8c44d

vasilytric committed:
[SYCL][ESIMD] Update log error message
